### ISSUE TITLE
Some Torchscript e2e tests have a mix of si64 and float operands

### ISSUE THREAD

**Issue by anupgangwar** [llvm/torch-mlir#532]
Hi,
In three tests 'ElementwiseSubScalarIntModule_basic', 'ElementwiseAddScalarIntModule_basic' and 'ElementwiseMulScalarModule_basic', there is a mix of si64 and float operands. I am wondering whether this is a valid test-case. See the Torch MLIR below:

```
module attributes {torch.debug_module_name = "ElementwiseSubScalarIntModule"}  {
  func @forward(%arg0: !torch.vtensor<[?,?],si64> loc(unknown)) -> !torch.vtensor<[?,?],f32> {
    %int2 = torch.constant.int 2 loc(#loc1)
    %float2.100000e00 = torch.constant.float 2.100000e+00 loc(#loc2)
    %0 = torch.aten.sub.Scalar %arg0, %float2.100000e00, %int2 : !torch.vtensor<[?,?],si64>, !torch.float, !torch.int -> !torch.vtensor<[?,?],f32> loc(#loc3)
    return %0 : !torch.vtensor<[?,?],f32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)

module attributes {torch.debug_module_name = "ElementwiseAddScalarIntModule"}  {
  func @forward(%arg0: !torch.vtensor<[?,?],si64> loc(unknown)) -> !torch.vtensor<[?,?],f32> {
    %float3.000000e00 = torch.constant.float 3.000000e+00 loc(#loc1)
    %int1 = torch.constant.int 1 loc(#loc0)
    %0 = torch.aten.add.Scalar %arg0, %float3.000000e00, %int1 : !torch.vtensor<[?,?],si64>, !torch.float, !torch.int -> !torch.vtensor<[?,?],f32> loc(#loc2)
    return %0 : !torch.vtensor<[?,?],f32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)

module attributes {torch.debug_module_name = "ElementwiseMulScalarModule"}  {
  func @forward(%arg0: !torch.vtensor<[?,?],si64> loc(unknown)) -> !torch.vtensor<[?,?],f32> {
    %float8.000000e00 = torch.constant.float 8.000000e+00 loc(#loc1)
    %0 = torch.aten.mul.Scalar %arg0, %float8.000000e00 : !torch.vtensor<[?,?],si64>, !torch.float -> !torch.vtensor<[?,?],f32> loc(#loc2)
    return %0 : !torch.vtensor<[?,?],f32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)

```

**Comment by anupgangwar** [comment]
Tagging @pashu123 and @Shukla-Gaurav since these tests were added by them. Tagging @ramiro050 and @silvasean for comments.

**Comment by pashu123** [comment]
These tests are eventually run by PyTorch here https://github.com/llvm/torch-mlir/blob/76728491cbefc8469a9a7bd467d783b5402461da/e2e_testing/torchscript/elementwise.py#L735 before generating the torchscript IR and hence these would be correct but I will wait for @ramiro050 @silvasean for the answer.

**Comment by ramiro050** [comment]
These tests, as mentioned by @Shukla-Gaurav, are specifically designed to test the type promotion logic. During the `refine-types` pass, the type promotion rules from PyTorch are used to determine the final type of the calculation, which for the tests mentioned happens here

https://github.com/llvm/torch-mlir/blob/76728491cbefc8469a9a7bd467d783b5402461da/lib/Dialect/Torch/Transforms/RefineTypes.cpp#L999

This is how `AtenMulScalarOp`, for example, knows that its result from an `si64` tensor and a float scalar is an `f32` tensor. In the case of `TorchToLinalg`, this information is then used to cast the operands to the right result type before performing the operation, as shown here:

https://github.com/llvm/torch-mlir/blob/76728491cbefc8469a9a7bd467d783b5402461da/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L2019-L2023

In general, as mentioned by @pashu123, because we are making a compiler for PyTorch, the source of truth for the behavior of the compiler is PyTorch itself. If it works in PyTorch, it is valid behavior:

```python
In [1]: import torch

In [2]: a = torch.randint(10, (2, 3))

In [3]: a.dtype
Out[3]: torch.int64

In [4]: result = torch.mul(a, 3.14)

In [5]: result.dtype
Out[5]: torch.float32
```

**Comment by anupgangwar** [comment]
Sounds good, I will keep the issue open till the support is baked in for any further questions on this topic.

**Comment by silvasean** [comment]
I think it would be ok to change the tests to i32, and then add at least one separate test specifically for i64 just so backends can track their support for that data type.

**Comment by anupgangwar** [comment]
Fixed in [PR #535](https://github.com/llvm/torch-mlir/pull/575).
